### PR TITLE
Inline into scan

### PR DESCRIPTION
**Pull request by havogt**
Adds
- A pass to inline non-trivial (non-SymRef) arguments of scans into the scanpass to make it possible to execute the scan without temporaries.
- A pass to remove `make_tuple(tuple_get(0,t),...)` patterns if tuple size `t` and the newly created tuple agree OR unconditionally if enabled. The unconditional version can be enabled if the tuple is unpacked within itir (we just don't use all elements), but not in the general case if the tuple is at the interface boundary. However, the gtfn backend allows assignment from larger to smaller tuples at the interface boundary. We could make this a feature and allow it also for embedded execution.
- A pass that applies the equivalent of an eta reduction even if order of parameters and arguments don't agree, but reshuffling parameters of the inner lambda (it does not work for functions).

### CODE REVIEW

**Comment by havogt** [comment]
For reference [f8d1d4f](https://github.com/GridTools/gt4py/pull/1122/commits/f8d1d4fe741ec19633b6c7b0895c214dde404592) was working with icon-exclaim solve nonhydro 52

File: src/functional/iterator/transforms/inline_into_scan.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+def _extract_symrefs(
+    nodes: list[ir.Expr], symtable: dict[eve.SymbolName, ir.Sym]
+) -> set[ir.SymRef]:
+    symrefs = []
+    for n in nodes:
+        if isinstance(n, ir.SymRef):
+            if _is_userdefined_symbolref(n, symtable):
+                symrefs.append(n)
+        else:
+            symrefs.extend(
```
Comment (havogt) [inline src/functional/iterator/transforms/inline_into_scan.py:34]: this doesn't work as we shouldn't look inside other scopes...

File: src/gt4py/next/iterator/transforms/merge_tuple.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+def _get_tuple_size(node: ir.Node) -> int:
+    infered_type = type_inference.infer(node)
+    assert isinstance(infered_type, type_inference.Val)
+    dtype = infered_type.dtype
+    assert isinstance(dtype, (type_inference.Tuple, type_inference.EmptyTuple))
+    return _count_tuple(0, dtype)
+
+
+@dataclass(frozen=True)
+class MergeTuple(eve.NodeTranslator):
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/merge_tuple.py:37]: Maybe `CollapseTuple`?

File: tests/next_tests/iterator_tests/test_with_toy_connectivity.py [hunk 3]
```diff
@@ -14,6 +14,19 @@
 
 import numpy as np
 import pytest
+from next_tests.toy_connectivity import (
```
Comment (havogt) [inline tests/next_tests/iterator_tests/test_with_toy_connectivity.py:17]: This is an unrelated change, that I thought I need. Should be extracted into a separate PR.

Comment (havogt) [reply tests/next_tests/iterator_tests/test_with_toy_connectivity.py:17]: Maybe if the reviewer agrees, we can keep it? There is the `functional` -> `next` commit inbetween and a simple `revert` + `cherry-pick` doesn't work...

**Review by tehrengruber — CHANGES_REQUESTED** [review]
First round

File: src/gt4py/next/iterator/transforms/extended_eta_reduction.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+def _is_scan(node: ir.Node):
+    return (
+        isinstance(node, ir.FunCall)
+        and isinstance(node.fun, ir.FunCall)
+        and node.fun.fun == ir.SymRef(id="scan")
+    )
+
+
+class ExtendedEtaReduction(NodeTranslator):
+    """Not strictly an eta reduction.
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/iterator/transforms/extended_eta_reduction.py]: done

File: src/gt4py/next/iterator/transforms/inline_into_scan.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    """
+
+    def visit_FunCall(self, node: ir.FunCall, **kwargs):
+        if _is_scan(node) and _should_inline(node):
+            original_scan_args = node.args
+            original_scan_call = node.fun
+            assert isinstance(original_scan_call, ir.FunCall)
+            refs_in_args = _extract_symrefs(original_scan_args, kwargs["symtable"])
+            original_scanpass = original_scan_call.args[0]
+            assert isinstance(original_scanpass, (ir.FunctionDefinition, ir.Lambda))
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/iterator/transforms/inline_into_scan.py]: done

File: src/gt4py/next/iterator/transforms/inline_into_scan.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            inlined,
+            opcount_preserving=True,
+            force_inline_lift=True,
+        )
+        if inlined == node:
+            break
+        node = inlined
+    else:
+        raise RuntimeError("Inlining lift and lambdas did not converge.")
+    return node
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/iterator/transforms/inline_into_scan.py]: done. checking now if it works with icon integration

[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/iterator/transforms/inline_into_scan.py]: looks good

File: tests/next_tests/ffront_tests/integration_tests/test_icon_like_scan.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+        test_setup.z_q_out,
+        offset_provider={"Koff": KDim},
+    )
+
+    assert np.allclose(test_setup.z_q_ref[:, 1:], test_setup.z_q_out[:, 1:])
+
+
+def test_solve_nonhydro_stencil_52_like_z_q_tup(test_setup, fieldview_backend):
+    # if fieldview_backend in [gtfn_cpu.run_gtfn, gtfn_cpu.run_gtfn_imperative]:
+    #     pytest.skip("Needs implementation of scan projector.")
```
Comment (tehrengruber) [inline tests/next_tests/ffront_tests/integration_tests/test_icon_like_scan.py:233]:
```suggestion
```

File: tests/next_tests/iterator_tests/test_scan.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    np_as_located_field,
+)
+from gt4py.next.iterator.runtime import fundef, offset
+from gt4py.next.program_processors.codegens.gtfn import gtfn_backend
+from gt4py.next.program_processors.runners import gtfn_cpu, roundtrip
+
+from .conftest import run_processor
+
+
+def test_scan_in_field_op(program_processor, lift_mode):
```
Comment (tehrengruber) [inline tests/next_tests/iterator_tests/test_scan.py:31]:
```suggestion
def test_scan_in_stencil(program_processor, lift_mode):
```

File: tests/next_tests/iterator_tests/test_scan_eta_reduction.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+def test_scan_eta_reduction():
+    testee = ir.Lambda(
+        params=[ir.Sym(id="x"), ir.Sym(id="y")],
+        expr=ir.FunCall(
+            fun=_make_scan("param_y", "param_x"),
+            args=[ir.SymRef(id="y"), ir.SymRef(id="x")],
+        ),
+    )
+    print(testee)
```
Comment (tehrengruber) [inline tests/next_tests/iterator_tests/test_scan_eta_reduction.py:41]:
```suggestion
```

File: tests/next_tests/iterator_tests/test_scan_eta_reduction.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    expected = _make_scan("param_x", "param_y")
+    print(expected)
```
Comment (tehrengruber) [inline tests/next_tests/iterator_tests/test_scan_eta_reduction.py:43]:
```suggestion
```

File: tests/next_tests/iterator_tests/test_symbol_ref_utils.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        ],
+    )
+    testee = _GetRealItirSymTable.apply(ir)
+    actual = get_user_defined_symbols(testee)
+    assert actual == {"target_symbol"}
+
+
+def test_collect_symbol_refs():
+    ...
+    # TODO
```
Comment (tehrengruber) [inline tests/next_tests/iterator_tests/test_symbol_ref_utils.py:60]:
```suggestion
def test_collect_symbol_refs():
    ...
    # TODO: Test collect_symbol_refs
```

File: tests/next_tests/iterator_tests/test_symbol_ref_utils.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        closures=[
+            itir.StencilClosure(
+                domain=itir.FunCall(fun=itir.SymRef(id="cartesian_domain"), args=[]),
+                stencil=itir.SymRef(id="deref"),
+                output=itir.SymRef(id="target_symbol"),
+                inputs=[],
+            )
+        ],
+    )
+    testee = _GetRealItirSymTable.apply(ir)
```
Comment (tehrengruber) [inline tests/next_tests/iterator_tests/test_symbol_ref_utils.py:53]:
```suggestion
    testee = GetSymtableForSymref.apply(ir, itir.SymRef(id="target_symbol"))
```
Just a suggestion, but it would be easier to digest if the symref you want the symtable for is explicitly given. Is the `_GetRealItirSymTable` actually needed or can't you just use `ir.annex["symtable"]`?

**Review by tehrengruber — CHANGES_REQUESTED** [review]
Only small things left.

File: tests/next_tests/iterator_tests/test_scan_eta_reduction.py (hunk 10 above)
Comment (havogt) [reply tests/next_tests/iterator_tests/test_scan_eta_reduction.py:43]: ups, sorry

File: src/gt4py/next/iterator/transforms/collapse_tuple.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+from gt4py import eve
+from gt4py.next.iterator import ir, type_inference
+
+
+def _count_tuple(count: int, dtype: type_inference.Tuple | type_inference.EmptyTuple) -> int:
+    if isinstance(dtype, type_inference.EmptyTuple):
+        return count
+    assert isinstance(dtype.others, (type_inference.Tuple, type_inference.EmptyTuple))
+    return _count_tuple(count + 1, dtype.others)
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/iterator/transforms/collapse_tuple.py:25]: added to Tuple

File: tests/next_tests/iterator_tests/test_symbol_ref_utils.py (hunk 12 above)
Comment (havogt) [reply tests/next_tests/iterator_tests/test_symbol_ref_utils.py:53]: I want to test that this pass extracts _only_ the user defined symrefs, therefore I want the real symbol table as a pass would give you (this symtable will contain the builtins). The symtable annex should not be used outside of a pass as recently discussed here https://github.com/GridTools/gt4py/pull/1137/files. So I am not sure if you didn't understand my test or if I didn't understand your comment.

Comment (havogt) [reply tests/next_tests/iterator_tests/test_symbol_ref_utils.py]: fixed

**Review by tehrengruber — APPROVED** [review]
